yarn/src/main/scala/org/apache/spark/deploy/yarn/ContainerPlacementStrategy.scala

```diff
@@ -0,0 +1,200 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.yarn
+
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.yarn.util.RackResolver
+
+import org.apache.spark.SparkConf
+
+private[yarn] trait ContainerPlacementStrategy {
+  type Locality = Array[String]
```

Comment:
If I understand correctly, a pair of localities is used for a single ContainerRequest.

If this is the case, it might be most clear to remove the locality type, and replace it with a ContainerLocalityPreferences class that has two fields: `nodes: Array[String]` and `racks: Array[String]`.  Then `localityOfRequestedContainers` could just return an Array of these.